src/features/releases/hooks/restrict-release-deletion.ts

```diff
@@ -0,0 +1,65 @@
+import { sbvrUtils, hooks, errors } from '@balena/pinejs';
+const { BadRequestError } = errors;
+
+hooks.addPureHook('DELETE', 'resin', 'release', {
+	PRERUN: async (args) => {
+		const { api } = args;
+		const affectedIds = await sbvrUtils.getAffectedIds(args);
+		if (affectedIds.length === 0) {
+			return;
+		}
+
+		const [applicationPinnedToRelease] = await api.get({
+			resource: 'application',
+			options: {
+				$top: 1,
+				$select: ['id', 'is_of__class'],
+				$expand: { should_be_running__release: { $select: 'raw_version' } },
+				$filter: {
+					should_be_running__release: {
+						$any: {
+							$alias: 'sbrr',
+							$expr: {
+								sbrr: {
+									id: { $in: affectedIds },
+								},
+							},
+						},
+					},
+				},
+			},
+		});
+
+		if (applicationPinnedToRelease != null) {
+			throw new BadRequestError(
+				`Unable to delete release ${applicationPinnedToRelease.should_be_running__release[0].raw_version} because it is the ${applicationPinnedToRelease.is_of__class}'s target release.`,
+			);
+		}
+
+		const [devicesPinnedToRelease] = await api.get({
+			resource: 'device',
+			options: {
+				$top: 1,
+				$select: 'uuid',
```

Comment:
If you only care for existence then it's likely better to select only the primary key, ie
```suggestion
				$select: 'id',
```